Tensile/SolutionStructs.py

```diff
@@ -2381,23 +2336,38 @@ def isDirectToVgprDoable(state, tc):
       reject(state, "DirectToVgpr%c does not supports WaveSeparateGlobalRead%c"%(tc, tc))
       return False
 
-    # Does not work with NumLoadsCoalesced>1
-    if state["NumLoadsCoalesced%c"%tc] > 1:
-      MIindex = 0 if tc == 'A' else 1
-      if state["NumLoadsCoalesced%c"%tc] > state['MIWaveTile'][MIindex] / state["GlobalReadVectorWidth"]:
-        reject(state, "DirectToVgpr%c does not supports NumLoadsCoalesced%c > MIWaveTile[%u] / GlobalReadVectorWidth"%(tc, tc, MIindex))
+    # Does not work with NumLoadsCoalesced != MIWaveTile / GlobalLoadVectorWidth
+    # (only for FractionalLoad = False)
+    if state["FractionalLoad"] == False:
+      if state["NumLoadsCoalesced%c"%tc] != state['MIWaveTile'][MIindex] / state["GlobalLoadVectorWidth%c"%tc]:
+        reject(state, "DirectToVgpr%c does not supports NumLoadsCoalesced%c(=%u) != MIWaveTile[%u](=%u) / GlobalLoadVectorWidth%c(=%u)"\
+                       %(tc, tc, state["NumLoadsCoalesced%c"%tc], MIindex, state['MIWaveTile'][MIindex], tc, state["GlobalLoadVectorWidth%c"%tc]))
         return False
+    # Does not work with MIWaveTile < GlobalLoadVectorWidth
+    if state['MIWaveTile'][MIindex] < state["GlobalLoadVectorWidth%c"%tc]:
+      reject(state, "DirectToVgpr%c does not supports MIWaveTile[%u](=%u) < GlobalLoadVectorWidth%c(=%u)"\
+                     %(tc, MIindex, state['MIWaveTile'][MIindex], tc, state["GlobalLoadVectorWidth%c"%tc]))
+      return False
 
     # Does not work with ExpandPointerSwap = False
     if not state["ExpandPointerSwap"]:
       reject(state, "DirectToVgpr%c does not supports ExpandPointerSwap = False"%(tc))
       return False
 
     # Does not work with VectorWidth != GlobalReadVectorWidth (VW = 2 + GRVW = 1 or VW = 1 + GRVW = 2 does not work)
-    if state["VectorWidth"] != state["GlobalReadVectorWidth"]:
+    if state["VectorWidth"] != state["GlobalLoadVectorWidth%c"%tc]:
       reject(state, "DirectToVgpr%c does not supports VectorWidth != GlobalReadVectorWidth"%(tc))
       return False
 
+    # Does not work with AssertFree1ElementMultiple % VectorWidth != 0 (edge shift case) for B
+    if tc == 'B' and state["AssertFree1ElementMultiple"] % state["VectorWidth"] != 0:
+      reject(state, "DirectToVgpr%c does not supports AssertFree1ElementMultiple %% VectorWidth != 0"%(tc))
+      return False
+
+    if kernel["ThreadSeparateGlobalRead%s"%tc]:
```

Comment:
Please fix lint error here.
  Tensile/SolutionStructs.py:2367:8: F821 undefined name 'kernel'

kernel should be state.